builds/jazz-build-module/azure-function-deploy-module.groovy

```diff
@@ -0,0 +1,308 @@
+#!groovy?
+import groovy.transform.Field
+/*
+* azure deployment module
+*/
+
+@Field def configLoader
+@Field def azureUtil
+@Field def utilModule
+@Field def scmModule
+@Field def events
+
+echo "azure deployment module loaded successfully"
+
+def initialize(configLoader, utilModule, scmModule, events, azureUtil){
+  this.configLoader = configLoader
+  this.utilModule = utilModule
+  this.scmModule = scmModule
+  this.events = events
+  this.azureUtil = azureUtil
+
+}
+
+
+def createFunction(serviceInfo){
+  azureUtil.setAzureVar(serviceInfo)
+  loadAzureConfig(serviceInfo)
+
+  def masterKey = invokeAzureCreation(serviceInfo)
+
+  def endpoint = "https://${serviceInfo.stackName}.azurewebsites.net/admin/functions/${serviceInfo.stackName}?code=$masterKey"
+  return endpoint
+
+}
+
+
+def sendAssetCompletedEvent(serviceInfo, assetList) {
+
+  def data = azureUtil.getAzureRequestPayload(serviceInfo)
+  data.tagName = 'service'
+  data.tagValue = serviceInfo.stackName
+  def output = azureUtil.invokeAzureService(data, "getResourcesByServiceName")
+
+  for (item in output.data.result) {
+    def assetType = item.kind
+    def id = item.id
+    switch (item.type) {
+      case "Microsoft.Storage/storageAccounts":
+        assetType = "storage_account"
+        break
+      case "Microsoft.ServiceBus/namespaces":
+        assetType = "servicebus_namespace"
+        break
+      case "Microsoft.EventHub/namespaces":
+        assetType = "eventhubs_namespace"
+        break
+      case "Microsoft.DocumentDB/databaseAccounts":
+        assetType = "cosmosdb_account"
+        break
+    }
+
+    events.sendCompletedEvent('CREATE_ASSET', null, utilModule.generateAssetMap(serviceInfo.serviceCatalog['platform'], id, assetType, serviceInfo.serviceCatalog), serviceInfo.envId)
+
+  }
+
+  if (assetList) {
+    for (item in assetList) {
+      def id = item.azureResourceId
+      def assetType = item.type
+      events.sendCompletedEvent('CREATE_ASSET', null, utilModule.generateAssetMap(serviceInfo.serviceCatalog['platform'], id, assetType, serviceInfo.serviceCatalog), serviceInfo.envId)
```

Comment:
We are sending CREATE_ASSET event twice in this function. Is this expected?

Reply:
yes.  
The first loop return us the assets created by this service.  The end loop is assets created by this service but not an azure resource such as database, table, queue, blob.
The loop assetList is track by this jenkins file but jenkins file does not know an azure resource is newly created or an existing which is to cover the requirement of allowing user enter existing resources.